### PR TITLE
Autograph 2.0: Sign addons

### PR DESCRIPTION
**Pull request by jvehent**
This set of patches will support signing Mozilla addons and web extensions in autograph. It does not rely on an HSM and performs all cryptographic operations in software.

The goal is to have autograph hold the private keys of intermediate CAs and use those to sign ephemeral addons certs.
- [x] Document addons signature format
- [x] Refactor autograph around a set of signer interfaces
- [x] Sign `sf` files with PKCS7 S/MIME detached signatures
- [x] Generate and sign ephemeral keypairs
- [x] Monitor signers expiration
- [x] Fix unit tests
- [x] Use OpenSSL pkcs7 in CI to verify test signatures
- [x] Integrate sops to read encrypted confs
- [x] Implement [audit logs](https://github.com/mozilla-services/autograph/issues/40)
- [x] Rewrite endpoints documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -476,3 +509,55 @@ Finally, digital signature helps us use Content Delivery Network without
...
+## Addon Signing (XPI)
+
+### Algorithm
+
+Addons are XPI files (zip archives) where each file in the archives is hashed
+with SHA-1 and MD5, and the manifest containing all the hashes is signed. To
+request a signature, the client must first compute the manifest file (stored in
+`META-INF/manifest.mf` in the XPI), then make a signature file (stored in
+`META-INF/mozilla.sf`) and submit the `sf` file to the signing API, alongside
+the addon ID (typically an email address). The signing API returns an S/MIME
```
Comment (kumar303) [inline README.md]: The addon ID looks like an email address but it rarely is a real email. Developers tend to use `any-kind-of-id` at their official add-on domain. Example: `public-release@ublock.org`.

File: README.md [hunk 2]
```diff
@@ -476,3 +509,55 @@ Finally, digital signature helps us use Content Delivery Network without
...
+detached signature the client must store under `META-INF/mozilla.rsa`.
+
+The signature file `mozilla.sf` simply contains md5.sha1 hashes of `manifest.mf`.
+(eg. `openssl dgst -binary -sha1 manifest.mf | base64`)
+
+To compute the signature, the signing API first generates an RSA key pair for
+the addon, called "ephemeral" because thrown away after signature. The certificate
```
Comment (kumar303) [inline README.md]: typo: because [it's] thrown away after [signing]

File: README.md [hunk 3]
```diff
@@ -476,3 +509,55 @@ Finally, digital signature helps us use Content Delivery Network without
...
+containing the public ephemeral key is signed by an intermediate issuer, itself
+signed by the AMO Root CA trusted in Firefox.
+
+The addon's certificate must have the following properties
+* SHA-256 signed
+* RSA public key of 2048 bits
+* subject CN set to the addon identifier (addon_id in the POST request)
+* 10 years validity
```
Comment (kumar303) [inline README.md]: Just a heads up, we once had a chemspill bug around expirations that weren't set right. I can't find it in bugzilla right now though.

**Comment by jvehent** [comment]
See #47

**Pull request merged**